### PR TITLE
Monitor input devices for keyboard attached state

### PR DESCRIPTION
**Pull request by MerlijnWajer**
Per commit message, monitor /dev/input for any additions or removals, and update keyboard attached logic mindlessly on any changes. (That is: don't keep track of what appears and disappears, just re-read them, there aren't that many of them anyway)

No files are kept open, other than the GIO file and file (directory) monitor. I didn't see any cleanup code (closing and deferencing on exit) - so I decided not to include that here either.

Also, move pkg-config calls to `configure.ac` instead of the `src/Makefile.am`.

### CODE REVIEW

**Comment by MerlijnWajer** [comment]
Also notice that the source file indentation is not consistent. I am inclined to fix that with `indent -linux`, but didn't and instead just used whatever the "local" indentation was.

File: configure.ac [hunk 1]
```diff
@@ -3,4 +3,24 @@ AM_INIT_AUTOMAKE(ke-recv-extra, 0.2)
 AM_CONFIG_HEADER(config.h)
 AC_PROG_CC
 
+PKG_CHECK_MODULES(GCONF, gconf-2.0)
+AC_SUBST(GCONF_CFLAGS)
+AC_SUBST(GCONF_LIBS)
```
Comment (spinal84) [inline configure.ac]: It's done done automatically, these two lines may be removed

Comment (MerlijnWajer) [reply configure.ac]: It took me a bit of searching, but I found this:

``` 
On older versions of pkg-config, the macro will not call AC_SUBST on these variables; modern versions (at least version 0.24) will take care of that already. Running it twice, though, will not cause problems, so if you have doubts, you should add a snippet like the following. 
```

From: https://autotools.io/pkgconfig/pkg_check_modules.html

I've think we have this almost everywhere. I'm OK with removing it here, if you think it makes sense.

Comment (spinal84) [reply configure.ac]: From Debian Stretch:
```
pkg-config --version
0.29
```

From https://cgit.freedesktop.org/pkg-config/refs/:
`pkg-config-0.24 ... 10 years`

It's there for at least 10 years. Maybe it's time to start using "modern" functionality?

But if you have doubts, ok

File: src/events.c [hunk 2]
```diff
@@ -41,6 +45,35 @@ static void inform_keyboard_attached(gboolean value)
         }
 }
 
+static void dir_changed_cb(GFileMonitor *monitor,
+               GFile *file, GFile *other_file,
+               GFileMonitorEvent event_type, gpointer user_data) {
+    (void)monitor;
+    (void)other_file;
+    (void)user_data;
```
Comment (spinal84) [inline src/events.c]: Why is this needed?

Comment (MerlijnWajer) [reply src/events.c]: Often, C compilers will complain about unused arguments passed to a function. Putting `(void)` in front of the argument silences these warnings. So if we would go for `-Wall -Wextra -Werror`, not having this code would cause problems.

File: src/events.c [hunk 3]
```diff
@@ -41,6 +45,35 @@ static void inform_keyboard_attached(gboolean value)
         }
 }
 
+static void dir_changed_cb(GFileMonitor *monitor,
+               GFile *file, GFile *other_file,
+               GFileMonitorEvent event_type, gpointer user_data) {
```
Comment (spinal84) [inline src/events.c]: I'd indent it like this:
```
static void dir_changed_cb(GFileMonitor *monitor,
                           GFile *file,
                           GFile *other_file,
                           GFileMonitorEvent event_type,
                           gpointer user_data)
{
```

File: src/events.c [hunk 4]
```diff
@@ -94,18 +131,13 @@ static void handle_e_bgkill_off_signal()
...
+static void handle_e_input_device_changed(void)
 {
-        if (get_input_device_state() != INPUT_DEVICE_ATTACHED) {
-               inform_keyboard_attached(FALSE);
-        }
+        reread_input_device_state();
+        inform_keyboard_attached(get_input_device_state() == INPUT_DEVICE_ATTACHED);
 }
 
+
```
Comment (spinal84) [inline src/events.c]: Is this extra line here on purpose or by chance?

File: configure.ac (hunk 1 above)
Comment (MerlijnWajer) [reply configure.ac]: Done in a separate commit (I didn't want to touch both the commits)

File: src/events.c (hunk 4 above)
Comment (MerlijnWajer) [reply src/events.c]: Removed in new push

File: src/events.c (hunk 2 above)
Comment (spinal84) [reply src/events.c]: Then I don't understand why you didn't put all unused arguments here. I counted five.
If the compiler doesn't complain, I'd better delete this, but it's up to you.

Comment (MerlijnWajer) [reply src/events.c]: Because I copied that part of the code from MCE, which does use the other variables. Oh, how copy-paste introduces errors. :) Will fix.

File: src/events.c [hunk 5]
```diff
@@ -41,6 +45,37 @@ static void inform_keyboard_attached(gboolean value)
...
+    (void)monitor;
+    (void)file;
+    (void)other_file;
+    (void)event_type;
+    (void)user_data;
+
+    handle_event(E_INPUT_DEVICE_CHANGED);
+}
+
+static int init_event_monitor() {
```
Comment (spinal84) [inline src/events.c]: Curly brace after function declaration goes on newline everywhere in this file, except this line.

File: src/events.c (hunk 3 above)
Comment (MerlijnWajer) [reply src/events.c]: I'm inclined to keep it the way it is, unless you have strong objections.

Comment (spinal84) [reply src/events.c]: The only objection is that such indentation is used everywhere (in this package) for multiline arguments. But it's up to you.

Comment (spinal84) [reply src/events.c]: Examples:
```
static gboolean match_event_file_by_caps(const gchar *const filename,
                                         const int *const ev_types,
                                         const int *const ev_keys[])
{
```
```
static
DBusHandlerResult sig_handler(DBusConnection *c, DBusMessage *m,
                              void *data)
{
```

File: src/events.c [hunk 6]
```diff
@@ -41,6 +45,37 @@ static void inform_keyboard_attached(gboolean value)
...
 }
 
+static void dir_changed_cb(GFileMonitor *monitor,
+               GFile *file, GFile *other_file,
+               GFileMonitorEvent event_type, gpointer user_data) {
+    (void)monitor;
+    (void)file;
+    (void)other_file;
+    (void)event_type;
+    (void)user_data;
```
Comment (spinal84) [inline src/events.c]: There're some examples of unused variables in the sources. In all cases the unused variables don't have "(void)expression;"
```
static gboolean
kdbus_sock_read_cb(GIOChannel* ch, GIOCondition cond, gpointer not_used)
```
```
static gboolean
kdbus_sock_event_cb(GIOChannel *ch, GIOCondition cond, gpointer not_used)
```
`void handle_kevent(DBusMessage* m)`

All examples are taken from this package (ke-recv-extra)

File: src/events.c (hunk 3 above)
Comment (MerlijnWajer) [reply src/events.c]: Ok, will fix.

File: src/events.c (hunk 5 above)
Comment (MerlijnWajer) [reply src/events.c]: Will fix.

**Comment by MerlijnWajer** [comment]
@spinal84 - I've changes all the things you mentioned I think. Is it good to go now?

**Comment by spinal84** [comment]
Don't see the changes.

File: src/events.c [hunk 7]
```diff
@@ -49,6 +84,10 @@ static void do_global_init()
         init_bgkill_state();
         init_input_device_state();
 
+        if (init_event_monitor()) {
```
Comment (spinal84) [inline src/events.c]: It's better to use opposite here: `if (! init_event_monitor())`
It's because in C programs the return result is usually implied by boolean, which mean 0 = FALSE.
When a programmer sees do_global_init() returns something, he would expect to see TRUE or FALSE (1 or 0).
TRUE means OK, FALSE means FAIL or ERROR.

Comment (MerlijnWajer) [reply src/events.c]: Yes and no. Yes, boolean logic works that way, but it's very common for C functions to return `0` on no error.

Comment (spinal84) [reply src/events.c]: What I mean:
```
        if (! init_event_monitor()) {
            ULOG_WARN_F("Failed to init event monitor");
        }
```
... instead of:
```
        if (init_event_monitor()) {
            ULOG_WARN_F("Failed to init event monitor");
        }
```

Comment (spinal84) [reply src/events.c]: The exception to this rule is main() function which returns 0 on success.

Comment (spinal84) [reply src/events.c]: https://www.quora.com/What-is-the-difference-between-return-0-and-return-1

Comment (spinal84) [reply src/events.c]: Which function on your mind returns 0 on success?

Comment (MerlijnWajer) [reply src/events.c]: Random pick, read:

```
RETURN VALUE
       On success, the number of bytes read is returned (zero indicates end  of  file),
       and  the  file  position is advanced by this number.  It is not an error if this
       number is smaller than the number of bytes requested; this may happen for  exam‐
       ple  because fewer bytes are actually available right now (maybe because we were
       close to end-of-file, or because we are reading from a pipe, or  from  a  termi‐
       nal), or because read() was interrupted by a signal.  See also NOTES.

       On  error,  -1 is returned, and errno is set appropriately.  In this case, it is
       left unspecified whether the file position (if any) changes.
```

stat:

```
RETURN VALUE
       On success, zero is returned.  On error, -1 is returned, and errno is set appro‐
       priately.
```

This is also true for most linux system calls, glibc calls, etc.

Comment (spinal84) [reply src/events.c]: Which function is that?
I know some low level functions can return 0 on success and <0 on error. The most functions which set errno probably.
But it's not our case here.

Comment (MerlijnWajer) [reply src/events.c]: Also: `chown`, and also:

```
Originally the return value would have corresponded to an error code. Imagine there's a big table of all the possible erroneous outcomes a program may have. The program terminates; you want to know what, if anything, went wrong. If a non-zero integer were returned, you'd look it up in your table of error codes. If 0 is returned, it's equivalent to "nothing went wrong."
```

Also:

```
This is a remnant of the C/Unix legacy.

If your program is being called within a script, is is useful for the script to know whether your program was successful. Thus, was born the convention of returning 0 for success from executable programs.

C did not have native boolean types, so this was an integer. And this allowed us to return different numbers for different kinds of errors.

This convention then spread to regular functions, since, as others have posted, C/C++ did not originally have exceptions, and even once we had exceptions, it took discipline to use them correctly, and there was a lot of legacy code (and legacy programmers) using the return value convention.
```

https://stackoverflow.com/questions/2370986/what-does-returning-zero-by-convention-mean

Comment (spinal84) [reply src/events.c]: If you are talking about read() function, it can't return 1 on success, because it returns the number of bytes read. Actually, it returns 1 if one byte was read. So the convention above can't be used there.

Comment (MerlijnWajer) [reply src/events.c]: Just go through `man 2 syscall` and you will see many of them return 0 on success. I'm happy to change other parts of the code, but I'm not going to change it because you happen to not agree with my pick for return value. The function is only called once, it's an init function. ;)

Comment (spinal84) [reply src/events.c]: Your function doesn't return -1 on error and it doesn't set errno. It only returns two values, 0 on success, 1 on errror. Which a programmer will read as "FALSE on success, TRUE on error". Do you know other functions which do the same?

Comment (spinal84) [reply src/events.c]: Anyway, it's your code, it's your choice. I just give you my feedback.

Comment (MerlijnWajer) [reply src/events.c]: Alright. I'll leave it like this. If the function ever needs to init more then there are other error codes besides `1` that can be returned. Thanks for the review!

File: src/events.c [hunk 8]
```diff
@@ -41,6 +45,36 @@ static void inform_keyboard_attached(gboolean value)
...
+
+static int init_event_monitor()
+{
+    GError *error = NULL;
+
+    dev_input_gfp = g_file_new_for_path(DEV_INPUT_PATH);
+
+    if ((dev_input_gfmp = g_file_monitor_directory(dev_input_gfp,
+                               G_FILE_MONITOR_NONE,
+                               NULL, &error)) == NULL) {
```
Comment (spinal84) [inline src/events.c:64]: This expression looks a bit complicated, I'd rewrite to:
```
    dev_input_gfmp = g_file_monitor_directory(dev_input_gfp, 
                                              G_FILE_MONITOR_NONE, 
                                              NULL, &error);

    if (dev_input_gfmp == NULL) {
```

Comment (spinal84) [reply src/events.c:64]: Complicated code is harder for reading and supporting.

**Comment by MerlijnWajer** [comment]
I missed the last comment, sorry.
